pkg/ensurance/cm/cpumanager/cpu_manager.go

```diff
@@ -0,0 +1,641 @@
+	podInformer coreinformers.PodInformer,
+	nrtInformer topologyinformer.NodeResourceTopologyInformer,
+) (CPUManager, error) {
+	machineInfo, err := cadvisorManager.GetMachineInfo()
+	if err != nil {
+		return nil, err
+	}
+	topo, err := topology.Discover(machineInfo)
+	if err != nil {
+		return nil, err
+	}
+	klog.InfoS("Detected CPU topology", "topology", topo)
+
+	initialContainers := buildContainerMapFromRuntime(containerRuntime)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build map of initial containers from runtime: %v", err)
+	}
+
+	cm := &cpuManager{
+		nodeName:         nodeName,
+		workqueue:        workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "cpumanager"),
+		podLister:        podInformer.Lister(),
+		nrtLister:        nrtInformer.Lister(),
+		podSync:          podInformer.Informer().HasSynced,
+		nrtSync:          nrtInformer.Informer().HasSynced,
+		defaultCPUPolicy: defaultCPUPolicy,
+		reconcilePeriod:  reconcilePeriod,
+		lastUpdateState:  cpumanagerstate.NewMemoryState(),
+		containerRuntime: containerRuntime,
+		containerMap:     containermap.NewContainerMap(),
+	}
+
+	_ = podInformer.Informer().AddIndexers(cache.Indexers{
+		cpuPolicyKeyIndex: func(obj interface{}) ([]string, error) {
+			pod, ok := obj.(*corev1.Pod)
+			if !ok {
+				return []string{}, nil
+			}
+			policyName := cm.getPodCPUPolicyOrDefault(pod)
+			return []string{policyName}, nil
+		},
+	})
+
+	podIndexer := podInformer.Informer().GetIndexer()
+	getPodFunc := func(cpuPolicy string) ([]*corev1.Pod, error) {
+		objs, err := podIndexer.ByIndex(cpuPolicyKeyIndex, cpuPolicy)
+		if err != nil {
+			return nil, err
+		}
+		pods := make([]*corev1.Pod, 0, len(objs))
+		for _, obj := range objs {
+			pod, ok := obj.(*corev1.Pod)
+			if !ok {
+				continue
+			}
+			// Succeeded and failed pods are not considered because they don't occupy any resource.
+			// See https://github.com/kubernetes/kubernetes/blob/f61ed439882e34d9dad28b602afdc852feb2337a/pkg/scheduler/scheduler.go#L756-L763
+			if pod.Status.Phase != corev1.PodSucceeded && pod.Status.Phase != corev1.PodFailed {
+				pods = append(pods, pod)
+			}
+		}
+		return pods, nil
+	}
+
+	podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: cm.enqueuePod,
+	})
+
+	cm.activePods = func() ([]*corev1.Pod, error) {
+		allPods, err := cm.podLister.List(labels.Everything())
+		if err != nil {
+			return nil, err
+		}
+		activePods := make([]*corev1.Pod, 0, len(allPods))
+		for _, pod := range allPods {
+			if !utils.IsPodTerminated(pod) {
+				activePods = append(activePods, pod)
+			}
+		}
+		return activePods, nil
+	}
+
+	cm.policy = NewStaticPolicy(topo, getPodFunc)
+
+	cm.state, err = cpumanagerstate.NewCheckpointState(
+		stateFileDirectory,
+		cpuManagerStateFileName,
+		cm.policy.Name(),
+		initialContainers,
+	)
+	if err != nil {
+		klog.ErrorS(err, "Could not initialize checkpoint state, please drain node and remove policy state file")
```

Comment:
why drain node is required on error? the state file is inside the agent pod not on the host right?

Reply:
We mount kubelet root path into container, so the crane state file will be also created in host path. It seems like we don't need to drain node, I just copy the code from `k/k`. Let me make a change here.